### PR TITLE
[RDY] Extended testing framework + modeline tests

### PR DESCRIPTION
**Pull request by TheBB**
This is just me playing around with ideas for #2440, #110.

With this PR, to add tests to a layer, you have to do three things:

1. Make a subdirectory of `tests` corresponding to your layer, with some test files. (See `/tests/core` or `/tests/spacemacs` in this PR for examples).
2. Write an `init.el` in that directory as well. This should be a minimal `.spacemacs` file for enabling your layer. See `/tests/spacemacs/init.el` for the extreme case of nothing.
3. Write a `Makefile` in that directory that lists the files you just wrote. See `/tests/core/Makefile` and `/tests/spacemacs/Makefile` for examples. This should be literally nothing other than copying `/spacemacs/Makefile` (don't do the one from core), replacing the files with the actual files you wrote in step 1. You may have to add some more `../` to the `include` as well though. The `LOAD_FILES` are relative to `~/.emacs.d/` (and there should be a very good reason if this is not just `init.el`), while the `TEST_FILES` are relative to the test directory.
4. Add the layer to the list of tests in `travis/run_build.sh`.

To run tests, you should go into the test directory for the layer you want to test and execute `make test`. This will run the tests naïvely. That is, it just runs the setup you currently have on your computer, and it ignores the `init.el` file you wrote in step 2. If you don't have the layer enabled in your personal dotfile, for example, you should expect the tests to fail. This is done to prevent testing from making changes in the developer's `~/.emacs.d/elpa`. It's intended as a quickie mode compared to Travis.

Travis, on the other hand, will go through core as well as all layers (skipping those without a makefile) to run all tests. For each layer, it will run emacs once with the environment variable `SPACEMACSDIR` set to `/tests/<layer>/`. This should make it choose the right dotfile for each test. This should **not** be done in each makefile individually, because that would cause test runs locally to mess up the elpa folder!

Additionally, Travis will delete the `elpa` and `.cache` folders between each layer test. Is this necessary? Well… maybe not. Spacemacs should of course ensure that the relevant packages are available for all runs, even if you use a different dotfile. I could probably remove this if we see the CI builds take too long. Right now it takes just short of a minute to do the whole build as it stands on this branch, which includes:

* Cloning the repo
* Installing emacs
* Running the core tests
* Installing all the packages required by the spacemacs layer

**Notes:**

* ~~As it stands, this PR doesn't include any actual _tests_, except for a dummy that does basically nothing.~~
* I have added the `ci` branch to the Travis file for testing (the testing), but I can remove that when it's time to merge.
* Coverage would be cool. But later. :-D

Comments? I don't expect this to be the "done deal" but it's perhaps a decent starting point for discussion.

### CODE REVIEW

**Comment by robbyoconnor** [comment]
:+1: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart:

File: spacemacs/tests/tests.el [hunk 1]
```diff
@@ -0,0 +1,2 @@
+(ert-deftest a-random-test ()
```
Comment (syl20bnr) [inline spacemacs/tests/tests.el]: Is a `require` for the test `init.el` file needed here ?
I don't see the test `init.el` file loaded anywhere.

**Comment by syl20bnr** [comment]
I'm pretty convinced by your proposal :+1:

I have 3 remarks:

### Duplication
There is too much duplication in the makefiles, we can define a `spacemacs.mk` file at the root of the project to factorize the rules and then a test makefile could be only a set of variables. Another possibility is to make a unique `Makefile` parameterized with environment variables, these variables would the parameters of a script `test.sh`.

### File names
I don't like the name `init.el` in the tests, it's better to keep only one `init.el` in the emacs dotdirectory which is `~/.emacs.d/init.el`. A small renaming to `init-tests.el` could do it. Or we could imagine a `.spacemacs-tests` file too.

### Test files location
I usually like to have things grouped together but here I'm split between storing the tests in the layer/core directories and storing the tests in `~/.emacs.d/tests`. I feel that putting tests outside is slightly better for such a project. Having tests is really important for any kind of users but they should be made as transparent as possible for the end user, I don't feel that putting tests folder everywhere is the right thing to do.

**Comment by TheBB** [comment]
Thanks for your feedback, @syl20bnr. I have updated with respect to your first and third suggestion.

The `init.el` files are actually dotfiles, which I let Spacemacs find by setting the `SPACEMACSDIR` environment variable. AFAIK it only looks for `~/.spacemacs` and `$(SPACEMACSDIR)/init.el`, so that's why it has to have that name.

Of course, I could name it something else and copy it to `~/.spacemacs` before each test, too.

**Comment by Devagamster** [comment]
One minor concern is that make on windows is a tad odd. It might be a necessary evil but it probably won't be the most approachable of choices. Is it just the default because its used elsewhere in the project?

**Comment by syl20bnr** [comment]
I thought about that, it should work by installing Cygwin with make package
and a .bat file to drive it. At least I use this setup at work to build
Erlang applications on multiple platforms and it works OK.


-syl20bnr-

**Comment by Devagamster** [comment]
So long as its documented how to run the tests on windows it should be fine. Too many places say "run make" without a description of how to do so on windows and leave people less familiar with Cygwin out in the cold.

**Comment by robbyoconnor** [comment]
@Devagamster, for somebody using an editor like emacs, it's not too much to assume they know how to use cygwin.

**Comment by TheBB** [comment]
I changed the name of the dotfile, and added some modeline tests.

**Comment by justbur** [comment]
I had a slightly different goal in mind, but this is related to #2127

**Comment by robbyoconnor** [comment]
@justbur -- why not make a linter for `.spacemacs` -- rather than full blown tests?

**Comment by syl20bnr** [comment]
I like the idea of @justbur and flycheck now support custom checkers written in elisp so I also like the idea of @robbyoconnor. An example of a checker written in elisp can be found here: https://github.com/purcell/flycheck-package

**Comment by robbyoconnor** [comment]
It's *WAY* too heavy weight -- The testing framework is for the layers -- but for one single file...it seems like we're using too big a tool for such a small task.

**Comment by justbur** [comment]
I don't understand how the 5 short functions I wrote can be considered way too heavyweight. 

I think the more important consideration is that the dotfile tests need to be run by people who don't understand spacemacs and I think it's too much to have them setup flycheck, which is not included by default, when they are asking for help. Whatever the dotfile test is it should be a simple command that anyone can run and interpret the output of from a basic installation.

**Comment by robbyoconnor** [comment]
Now we gotta work on improving the tests for the various layers :D

**Comment by TheBB** [comment]
Yes, @syl20bnr, I think this is ready now. It doesn't include a whole lot of tests (just a few for the modeline, more to come later) but it gets the ball rolling, at least. I put a guildeline in `CONTRIBUTE.org`.

@justbur's dotfile tests are useful but a separate issue from this.

**Comment by TheBB** [comment]
@syl20bnr If you're willing to have a look at this I can rebase. There are no significant changes touching any real code, if you're concerned about release.

**Comment by robbyoconnor** [comment]
:grinning:

**Comment by syl20bnr** [comment]
@TheBB you can rebase it, I'll look into it during the week.

**Comment by TheBB** [comment]
Okay, that was easy enough. I guess the github 'can this be auto-merged' check is a bit conservative.

File: travis/run_build.sh [hunk 2]
```diff
@@ -10,6 +10,26 @@
 ## This file is not part of GNU Emacs.
 ##
 ## License: GPLv3
-ln -s `pwd` /home/travis/.emacs.d
-cd ~/.emacs.d/core
-make test
+
+tests=("core" "spacemacs")
+
+if [ $USER != "travis" ]; then
```
Comment (TheBB) [inline travis/run_build.sh]: If someone's username is actually travis, be careful. :-P

**Comment by robbyoconnor** [comment]
@TheBB -- maybe migrate to the docker-based infrastructure...everything we need to do for testing doesn't rely on being root...so we have no reason to have sudo.

**Comment by TheBB** [comment]
Yeah, last I checked there was a PR that had to be accepted at travis, to whitelist a package? Something like that... I don't remember where I read it, though.

**Comment by robbyoconnor** [comment]
@TheBB, go look at the travis build...you'll see a notice of us being on the legacy platform...click the link ;)

**Comment by syl20bnr** [comment]
Works like a charm, triple kudos @TheBB :100: 
Spacemacs quality level has just been raised by a good amount :+1: 

Cherry-picked into `develop`, [...] you branch. :-)

**Comment by TheBB** [comment]
@robbyoconnor I know... I'm referring to https://github.com/syl20bnr/spacemacs/issues/2512#issuecomment-127661484 and by extension https://github.com/travis-ci/apt-source-whitelist/pull/95.

**Comment by TheBB** [comment]
Thanks @syl20bnr!

**Pull request closed without merging**